### ISSUE TITLE
Command-line scripts added to homebridge-blinds

### ISSUE THREAD

**Issue by dxdc** [hjdhjd/homebridge-blinds-cmd#82]
@hjdhjd 

Finally got around to updating https://github.com/dxdc/homebridge-blinds to support command line scripts. Wondering if you may find it more feature rich, since your plugin is forked from the original homebridge-blinds repo?

Example config is as follows, just prefix your commands with `file://` to have them run directly instead of calls via http.

```json
        {
            "name": "TEST",
            "up_url": "file://python /var/lib/homebridge/Blinds_UP.py",
            "motion_up_time": 25000,
            "down_url": "file://python /var/lib/homebridge/Blinds_DOWN.py",
            "motion_down_time": 25000,
            "stop_url": "file://python /var/lib/homebridge/Blinds_Pause.py",
            "accessory": "BlindsHTTP"
        }
```

I spent quite a bit of time tuning the logic for HTTP, so you may find it beneficial for your use case. The only feature (I think?) that may not be supported in `homebridge-blinds` is your use of `refreshRate`. Anyhow, would be happy to take any feedback you have.

**Comment by hjdhjd** [comment]
@dxdc And try to honor the template and you won't get things autoclosed. 😄

**Comment by hjdhjd** [comment]
I'll take a closer look next week, but at initial glance, the very first thing that I get hit with is that it's incredibly complicated and overly feature rich. It's intimidating and nearly unnavigable unless you're going to spend serious time getting acquainted with the plugin. That's my quick take from a short glance. I'd suggest simplifying the options and breaking up the documentation into the various use cases / operating modes you're trying to support. It's just *a lot* to read through versus, here's a couple of parameters and I'm up and running.

For what it's worth...

**Comment by hjdhjd** [comment]
Also - you still have no way of effectively linking multiple shades together, or dealing with status changes that happen outside of homebridge-blinds. For my use cases these are hard requirements.

To support it, you need to add support for a status command which provides the canonical status of a blind. Without it, everything else falls apart quickly in a world where not everything is controlled in one place.

**Comment by dxdc** [comment]
Thanks, yes, the config has grown to be more complicated than I'd like. The majority of the features are optional. It's on the todo list for sure.

I need to redo the config - some of it is historical since I inherited the plugin and didn't want to break anyone's set up. And, a lot of the parameters are for people who have more complex HTTP requirements like custom headers, post/get, etc.

> you still have no way of effectively linking multiple shades together
How is this accomplished in your current set up?

> or dealing with status changes that happen outside of homebridge-blinds
Yes, there is support for this via `position_url` but no polling yet. HK has built-in polling, but need to add the ability to poll on some kind of timed interval like in your example.

**Comment by hjdhjd** [comment]
Look at the example scripts for my plugin to see how I accomplish this today. Essentially, I maintain the link relationship in scripts and have a separate data store that retains the position values, so when status is queried, the right thing is always brought back to HomeKit. This is also why it's *essential* to have a status script. That way you always maintain proper state.

One suggestion you may want to try...fork my code and reverse the features in your plugin into it. In my quick scan, it should be a straightforward process. My code should be pretty straightforward and well-commented (and TypeScript which you can view as a bonus or not, depending on your taste :smile:).

I'd be happy to retire my plugin, as long as you can support the full universe of functionality of mine and be a proper superset. Ideally, I'd love to be able to effectively just shift my plugin options into yours. :smile:

TL;DR: would love to retire my plugin if `homebridge-blinds`:

* support status reporting and polling properly.
* adjust plugin configuration to be more consistent and thought out.

Thoughts?

BTW - good hearing from you...been a while!

**Comment by dxdc** [comment]
thanks @hjdhjd and happy holidays to you!

I am confused regarding your "status" reporting, since that is already supported via `position_url`. The only thing missing is a periodic refresh.

> Setting position isn't the issue

Setting position can be accomplished via webhook (on a one-time basis)

> reporting the current position that's an unsolved edge case in homebridge-blinds

Didn't quite get this part. `position_url` is polling for status (really, "position" not "status", since it's a 0-100 response). However, it's doing so at a frequency defined by HK, not a periodic poll. So it may not update as frequently.

> adjust plugin configuration to be more consistent and thought out.

Sure, it's on the list. If you can share your existing config, I can convert yours for you for testing purposes.

**Comment by hjdhjd** [comment]
@dxdc I think we're talking past each other a bit. I need status / position reporting to be a script, not just a  webhook. Same with setting a position. Webhooks are an unnecessary complication in my use case, and most of how you've built `homebridge-blinds` assumes some type of an HTTP-centric API. I'm not always operating in that world, and it's important (to me) that support for command-line scripts be first-class citizens when it comes to functionality / capability.

I'd love to retire my plugin, but I need the full range of capabilities accessible via only command line scripts, if I choose.

As to polling and the very real unresolved edge case that somehow we aren't communicating about - it's essential that polling be an element of status reporting. If you want to call it position reporting, fine - we're getting caught up on semantics. :smile: At the end of the day, the plugin *should* be able to periodically query, if I want it to, to refresh it's internal view of blind state relative to what I know it to be. I don't want to have to write a separate cron job to call the webhook every few seconds to update it's state. :smile: The plugin should poll if I want it to.

Consider scenarios where the shades are moved independently of the plugin, but I happen to be able to keep track of that state myself, separately. If the blind is controlled outside of your plugin, my only recourse is to write a cron job to periodically call the webhook for each blind. Having the plugin poll regularly for state allows me to tell you "hey, the blind's real position is 70, you need to update your internal representation of state and report it  to HomeKit". That's essential. Without it, the plugin loses most of it's value for me.

And sure, technically HK decides when it wants to query state, but that's not what I'm saying. HK knows whatever state you tell it. If the state has changed *independently of HomeKit / homebridge*, you need to publish an update (updateCharacteristic) that informs HomeKit of what the true current position is so that if I do the following:

Set position to 100 through HomeKit / Homebridge <--- HK reports blind is fully open. Yay!
Set position to 20 outside of HomeKit / Homebridge <--- HK reports blind is fully open. FAIL!!!! It's not.

In my plugin, you can optionally run a script at regular intervals (say 5 seconds or 10 seconds) that will report back the canonical position and update HK if needed. So that this happens:

Set position to 100 through HomeKit / Homebridge <--- HK reports blind is fully open. Yay!
Set position to 20 outside of HomeKit / Homebridge <--- homebridge-blinds-cmd runs the status script every 5 seconds, picks this up, and updates currentPosition to HomeKit, correctly representing the state.

The closest I can get to this in yours, is to trigger the webhook on each external state change...which is a solution that doesn't elegantly scale when you have a dozen motorized blinds (aka me!). Running 12 web servers on 12 ports just to be able to deal with position information feels unnecessarily complex.

These edge cases are important to me. :smile:

My suggestion (that you should absolutely ignore if you want since some of these are truly philosophical differences) would be to rethink the way this plugin works:

* Refactor it into a platform plugin rather than accessory plugin so you can create some awareness across blinds and deal with more complex grouping scenarios / configurations.
* Standardize the API a bit more and make it more consistent throughout. If you're going to have a webhook, okay, but let me control / set multiple blinds through the same webhook - don't create a new webhook for every. single. blind.
* If you're going to support command-line scripts, great. Let me do everything you're enabling via webhooks and URLs with a command line script without the need for webhooks.

If you decide to take on any of the above...as I said earlier, you're welcome to fork my plugin and backport in your additions on the HTTP side of the house...it may help you streamline / simplify.

Happy holidays!

**Comment by dxdc** [comment]
@hjdhjd  Im still not sure that I get your point since a lot of the  features you're describing are already implemented.

> I need status / position reporting to be a script, not just a webhook. Same with setting a position.

Yes, this already exists. You can use a position_url (as HTTP) or as a script (via, `file://` prefix). This is already implemented and polled by HK natively. It does not however have a polling feature, which I will look into adding.

> Consider scenarios where the shades are moved independently of the plugin, but I happen to be able to keep track of that state myself, separately.

Above deals with this already. Once polling is added, it will refresh automatically. I'm also going to have it check for status before initiating a move command, which should really help avoid edge cases.

> In my plugin, you can optionally run a script at regular intervals (say 5 seconds or 10 seconds) that will report back the canonical position and update HK if needed. So that this happens:

Yes, polling with fix that. 

> The closest I can get to this in yours, is to trigger the webhook on each external state change...which is a solution that doesn't elegantly scale when you have a dozen motorized blinds (aka me!). Running 12 web servers on 12 ports just to be able to deal with position information feels unnecessarily complex.

No, this isn't needed. The only need for the webhook is for occasional updates that are more "push" based that don't reflect the position itself. For example, suppose you had a sensor when blind was fully open only, you could "push" this to the position. Otherwise, the webhook isn't needed.

> Refactor it into a platform plugin rather than accessory plugin so you can create some awareness across blinds and deal with more complex grouping scenarios / configurations.

I don't really see the advantage/difference of platform vs. accessory. There is no effect to grouping, to my knowledge? For scenes, you can just choose the accessories you wish to support. The only advantage I can see would be to have a single port for webhook, though not sure how important that is since it is an edge case for users as is. Still, will think about it if there is some other inherent advantage. The reason that platform makes less sense to me is because it's not as if we have a singular server which defines multiple devices. Each device must still be configured/defined separately.

> If you're going to support command-line scripts, great. Let me do everything you're enabling via webhooks and URLs with a command line script without the need for webhooks.

Command-line scripts are already supported, and webhooks aren't required. It will poll your status script... but probably not quite often enough.

Noted on the rest. Will think about it and update you. Thx for the discussion!

**Comment by hjdhjd** [comment]
Glad to hear that much of what I was advocating for is already there. That just means I've got to set some time aside to muck around with your plugin so I can have a more informed view. :smile:

Agree that polling will go a ways towards addressing some of the issues. If / when you choose to implement it, let me know. That's a substantial impetus for me.

The platform versus accessory bit is relevant if, like me, you would like to configure a whole bunch of shades in a single config block rather than duplicating them each and every time. It's be ideal if I could say, create a single "command template" like: http://some_ip/%s/up where %s is substituted for the shade name, or serial, or whatever unique ID you want to use. Granted, config pain is a one-time thing for most people. :smile:

The other reason is grouping. No, I'm not talking about scenes. Say I have two shades in a single room, but I actually choose to represent them as three different shades - one each, singularly, and a third shade allowing me to control both of them as a single entity. An accessory plugin only knows about that single accessory. A platform plugin knows about all the accessories residing on that platform. I can tell a platform plugin "hey, take these two shades and create a third, composite shade and properly maintain status across all three". This allows me to choose how I want to control things.

I can do this now on my own, if polling is implemented, and there's a clean command-line script API. I would need to maintain all that state information outside of the plugin, but it's just as easy to do it within the plugin. Creating composite shades (made up of different shades) are worthwhile.

Again - some of this is philosophical, some isn't. You've done a lot with the plugin. Look forward to seeing it mature further to meet the use cases I, and the users of my plugin encounter, and one day retiring it.

**Comment by dxdc** [comment]
> Agree that polling will go a ways towards addressing some of the issues. If / when you choose to implement it, let me know. That's a substantial impetus for me.

Will keep you posted. Working on a better TS port too. Just need to find the time :)

> The platform versus accessory bit is relevant if, like me, you would like to configure a whole bunch of shades in a single config block rather than duplicating them each and every time. It's be ideal if I could say, create a single "command template" like: http://some_ip/%s/up where %s is substituted for the shade name, or serial, or whatever unique ID you want to use. Granted, config pain is a one-time thing for most people. 😄

I see your point here. But, I also think that your use case may be very specific to your case alone. Then, it gets tricker because people will have multiple different kinds of templates with different blinds too.

> The other reason is grouping. No, I'm not talking about scenes. Say I have two shades in a single room, but I actually choose to represent them as three different shades - one each, singularly, and a third shade allowing me to control both of them as a single entity.

So, suppose shade 1 & 2 are grouped. Shade 1 is up, shade 2 is down. What is the status of your grouped control? But, regardless, grouping can still be done within HK itself using accessories... just create a 3rd blind and hook it to run whatever command you like. This also gets extremely complicated if those blinds had different heights so took different amounts of time to go up/down. Curious on your set up for that.

I have the same situation as you (e.g., side-by-side shades) but find that for my own use I just add them as a single composite accessory and don't also provide access to the individual blinds.

I think once polling is implemented it will be more sensible to give things a try and convert your config over. If you have an example config would love to see it so I can try to understand how to best address.
